light/client.go

```diff
@@ -644,20 +644,47 @@ func (c *Client) sequence(
 		err = VerifyAdjacent(c.chainID, trustedHeader, interimHeader, interimVals,
 			c.trustingPeriod, now, c.maxClockDrift)
 		if err != nil {
-			err := fmt.Errorf("verify adjacent from #%d to #%d failed: %w",
-				trustedHeader.Height, interimHeader.Height, err)
+			err := ErrVerificationFailed{From: trustedHeader.Height, To: interimHeader.Height, Reason: err}
 
 			switch errors.Unwrap(err).(type) {
 			case ErrInvalidHeader:
+				// If the target header is invalid, return immediately.
+				if err.To == newHeader.Height {
+					c.logger.Debug("Target header is invalid", "err", err)
+					return err
+				}
+
+				// If some intermediate header is invalid, replace the primary and try
+				// again.
 				c.logger.Error("primary sent invalid header -> replacing", "err", err)
 				replaceErr := c.replacePrimaryProvider()
 				if replaceErr != nil {
 					c.logger.Error("Can't replace primary", "err", replaceErr)
 					// return original error
 					return err
 				}
+
+				replacementHeader, replacementVals, fErr := c.signedHeaderAndValSetFromPrimary(newHeader.Height)
+				if fErr != nil {
+					c.logger.Error("Can't fetch header/vals from primary", "err", fErr)
+					// return original error
+					return err
```

Comment:
Do we want to throw the same `ErrVerificationFailed` errors for these situations as well - I guess it depends on what we consider the verification process

Reply:
I view replacement more as a helper process and more internal. Like the client, who's using the light client should not know we're performing replacement ideally. I.e. it's hidden from the client. The client only sees the verification errors.

Reply:
I see that makes sense 👍